nemo/core/app_state.py

```diff
@@ -0,0 +1,97 @@
+# ! /usr/bin/python
+# -*- coding: utf-8 -*-
+
+# =============================================================================
+# Copyright (c) 2020 NVIDIA. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+import threading
+
+# Sadly have to import this to avoid circular dependencies.
+import nemo
+
+
+class Singleton(type):
+    """ Implementation of a generic singleton meta-class. """
+
+    # List of instances - one per class.
+    __instances = {}
+    # Lock used for accessing the instance.
+    __lock = threading.Lock()
+
+    def __call__(cls, *args, **kwargs):
+        """ Returns singleton instance.A thread safe implementation. """
+        if cls not in cls.__instances:
+            # Enter critical section.
+            with cls.__lock:
+                # Check once again.
+                if cls not in cls.__instances:
+                    # Create a new object instance - one per class.
+                    cls.__instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        # Return the instance.
+        return cls.__instances[cls]
+
+
+class AppState(metaclass=Singleton):
+    """
+        Application state stores variables important from the point of view of execution of the NeMo application.
```

Comment:
I don't understand why this is needed? All epoch, device, etc. are in NeuralModuleFactory already